chain/client/src/chunk_inclusion_tracker.rs

```diff
@@ -1,37 +1,60 @@
+use chrono::{DateTime, Utc};
 use itertools::Itertools;
 use lru::LruCache;
+use near_primitives::checked_feature;
+use near_primitives::version::ProtocolVersion;
 use std::collections::HashMap;
+use std::sync::Arc;
 
+use near_chain_primitives::Error;
+use near_epoch_manager::EpochManagerAdapter;
+use near_primitives::block_body::ChunkEndorsementSignatures;
 use near_primitives::hash::CryptoHash;
-use near_primitives::sharding::ShardChunkHeader;
+use near_primitives::sharding::{ChunkHash, ShardChunkHeader};
 use near_primitives::types::{AccountId, EpochId, ShardId};
 
+use crate::chunk_validation::ChunkValidator;
 use crate::metrics;
 
 const CHUNK_HEADERS_FOR_INCLUSION_CACHE_SIZE: usize = 2048;
 const NUM_EPOCH_CHUNK_PRODUCERS_TO_KEEP_IN_BLOCKLIST: usize = 1000;
 
+struct ChunkInfo {
+    // chunk_header, received_time and chunk_producer are populated when we call mark_chunk_header_ready_for_inclusion
+    pub chunk_header: ShardChunkHeader,
+    pub received_time: DateTime<Utc>,
+    pub chunk_producer: AccountId,
+    // The signatures are populated later during call to chunk_validator.get_chunk_endorsement_signature
+    // This is a sort of cache to store previously fetched signatures.
+    pub signatures: Option<ChunkEndorsementSignatures>,
+}
+
 pub struct ChunkInclusionTracker {
     // Track chunks that are ready to be included in a block.
     // Key is the previous_block_hash as the chunk is created based on this block. It's possible that
     // the block included isn't of height previous_block_height + 1 in cases of skipped blocks etc.
-    // We store the map of chunks from [shard_id] to (chunk_header, chunk received_time, chunk_producer account_id)
-    prev_block_to_chunk_headers_ready_for_inclusion: LruCache<
-        CryptoHash,
-        HashMap<ShardId, (ShardChunkHeader, chrono::DateTime<chrono::Utc>, AccountId)>,
-    >,
+    // We store the map of chunks from [shard_id] to chunk_hash
```

Comment:
This is complex enough logic that I can't judge for sure whether it'll be bug-free. Is it practical to set up a local test for this?

Reply:
I'll see if it's possible to add some tests

Reply:
Actually... I'll follow up in a separate PR later